app/Http/Controllers/SupplyController.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+namespace Adshares\Adserver\Http\Controllers;
+
+use Adshares\Adserver\Models\NetworkCampaign;
+use Adshares\Adserver\Models\NetworkEventLog;
+use Adshares\Adserver\Models\EventLog;
+
+use Adshares\Adserver\Http\GzippedStreamedResponse;
+use Adshares\Adserver\Http\Utils;
+
+use Adshares\Adserver\Services\BannerFinder;
+
+use Adshares\Esc\Esc;
+
+use Illuminate\Http\Request;
+
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+
+// use Adshares\Services\Adselect;
+
+// TODO: review request headers // extract & organize ??
+
+/**
+ * HTTP api that is used by supply adserver to display banners and log relevant events
+ *
+ */
+class SupplyController extends Controller
+{
+    public function find(Request $request)
+    {
+        $response = new Response();
+
+        if ($request->headers->has("Origin")) {
+            $response->headers->set("Access-Control-Allow-Origin", $request->headers->get("Origin"));
+            $response->headers->set("Access-Control-Allow-Credentials", "true");
+            $response->headers->set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
+        }
+
+        if ($request->getRealMethod() == 'GET') {
+            $data = $request->getQueryString();
+        } elseif ($request->getRealMethod() == 'POST') {
+            $data = $request->getContent();
+        } elseif ($request->getRealMethod() == 'OPTIONS') {
+            $response->setStatusCode(204);
+            $response->headers->set('Access-Control-Max-Age', 1728000);
+            return $response;
+        }
+
+        $decoded = Utils::decodeZones($data);
+//         print_r($decoded);exit;
+        $zones = $decoded['zones'];
+
+        $tid = Utils::attachTrackingCookie(config('app.adserver_secret'), $request, $response, "", new \DateTime());
+
+        // use adselect here
+        $context = Utils::getImpressionContext($request, $data);
+
+        $impressionId = $decoded['page']['iid'];
+        if ($impressionId) {
+            $aduser_endpoint = config('app.aduser_endpoint');
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "app.aduser_endpoint" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)